### PR TITLE
chore: lint staged files before commit

### PR DESCRIPTION
**Pull request by InvictusMB**
Adds precommit hook to lint staged files. Also fixes prettier EOL handling on Windows.

Running linting on the whole project manually takes a while. It is much more convenient to lint only modified files automatically every time.

Or is there a good reason not to do so?

**EDIT:** Just saw #5462. This one is slightly different though.

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: .prettierrc [hunk 1]
```diff
@@ -3,6 +3,7 @@
   "singleQuote": true,
   "printWidth": 80,
   "trailingComma": "all",
+  "endOfLine": "auto",
```
Comment (bajtos) [inline .prettierrc:6]: This should go to a standalone commit. A new PR would be even better, if you don't mind.

Comment (InvictusMB) [reply .prettierrc]: Done in #5673

**Comment by InvictusMB** [comment]
@bajtos 
Updated the PR as requested. Also implemented bypassing `lint-staged` altogether via `LINT_STAGED=false`.

**Comment by InvictusMB** [comment]
Added remarks about pre-commit linting to the documentation.

File: .lintstagedrc [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+  "*.{ts,js}": [
+    "node packages/build/bin/run-prettier --write",
```
Comment (raymondfeng) [inline .lintstagedrc:3]: Can we only fix changed files for `prettier`?

File: .lintstagedrc [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+    "node packages/build/bin/run-eslint --fix --report-unused-disable-directives --cache"
+  ],
+  "*.{md}": [
+    "node packages/build/bin/run-prettier --write"
```
Comment (raymondfeng) [inline .lintstagedrc:7]: Ditto.

File: .lintstagedrc (hunk 2 above)
Comment (raymondfeng) [reply .lintstagedrc:3]: Can we also add a print out to describe how to skip the check if the command fails?

Comment (InvictusMB) [reply .lintstagedrc:3]: `lint-staged` runs command only on changed files that were staged for commit by definition.

This is the output when pre-commit hook fails:

<details>
  <summary>output</summary>
  
```
~ git commit -a -m "chore: fail prettier"
husky > pre-commit (node v12.16.0)
[STARTED] Preparing...
[SUCCESS] Preparing...
[STARTED] Running tasks...
[STARTED] Running tasks for *.{ts,js}
[STARTED] Running tasks for *.{md}
[STARTED] node packages/build/bin/run-prettier --write
[FAILED] node packages/build/bin/run-prettier --write [FAILED]
[FAILED] node packages/build/bin/run-prettier --write [FAILED]
[SUCCESS] Running tasks...
[STARTED] Applying modifications...
[STARTED] Reverting to original state because of errors...
[SUCCESS] Reverting to original state because of errors...
[STARTED] Cleaning up...
[SUCCESS] Cleaning up...

× node packages/build/bin/run-prettier --write:
[error] packages\repository\src\repositories\legacy-juggler-bridge.ts: SyntaxError: ';' expected. (464:8)
[error]   462 |   }
[error]   463 |
[error] > 464 |   fail here
[error]       |        ^
[error]   465 |
[error]   466 |   async updateById(
[error]   467 |     id: ID,
husky > pre-commit hook failed (add --no-verify to bypass)
``` 
</details>

Is this sufficient?

Comment (raymondfeng) [reply .lintstagedrc:3]: @InvictusMB I understand the command is only triggered for changes files. But `node packages/build/bin/run-prettier --write` itself scans all files.

Comment (raymondfeng) [reply .lintstagedrc:3]: I'm referring the possibility of https://github.com/strongloop/loopback-next/blob/63e8f6e1f91b0808696a11b123318d5089f2b2d0/lint-staged.config.js#L13.

Comment (raymondfeng) [reply .lintstagedrc:3]: The output is good enough to understand what's failing. My comment was to print out an option to skip the failure so that the pre-commit hooks can be skipped.

Comment (InvictusMB) [reply .lintstagedrc:3]: > I'm referring the possibility of
> 
> https://github.com/strongloop/loopback-next/blob/63e8f6e1f91b0808696a11b123318d5089f2b2d0/lint-staged.config.js#L13

Lint staged literally does that for us already.
That's what gets passed into `run-prettier`
```
run-prettier argv
[
  'C:\\DevTools\\nodejs\\node.exe',
  'D:\\Projects\\loopback-next\\packages\\build\\bin\\run-prettier',
  '--write',
  'D:/Projects/loopback-next/packages/build/bin/run-prettier.js',
  'D:/Projects/loopback-next/packages/repository/src/repositories/legacy-juggler-bridge.ts'
]
```

> The output is good enough to understand what's failing. My comment was to print out an option to skip the failure so that the pre-commit hooks can be skipped.

It is the last line in the output:
```
husky > pre-commit hook failed (add --no-verify to bypass)
```
So far there does not seem to be a way to customize that message.
If you mean highliting `LINT_STAGED=false` option, I'm not sure we should also promote a permanent switch here. It is described in the docs and visible in pre-commit hook definition in `package.json`.

File: package.json [hunk 4]
```diff
@@ -80,6 +82,7 @@
   "copyright.owner": "IBM Corp.",
   "husky": {
     "hooks": {
+      "pre-commit": "if-env LINT_STAGED=false && echo \"lint-staged disabled via LINT_STAGED env var\" || lint-staged",
```
Comment (bajtos) [inline package.json:85]: Let's use `0` instead of `false`, it's shorter to type :)

```suggestion
      "pre-commit": "if-env LINT_STAGED=0 && echo \"lint-staged disabled via LINT_STAGED env var\" || lint-staged",
```

File: .lintstagedrc [hunk 5]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (raymondfeng) [inline .lintstagedrc:1]: Can we switch to `lint-staged.config.js` as the default config for `lint-staged`. The JS style gives us more flexibility.

File: package.json (hunk 4 above)
Comment (InvictusMB) [reply package.json]: Updated

File: .lintstagedrc (hunk 5 above)
Comment (InvictusMB) [reply .lintstagedrc:1]: I'd rather not. But it's only my personal preference to introduce more flexibility when needed. Feel free to incorporate the changes from this PR into #5462.

**Comment by raymondfeng** [comment]
@InvictusMB Thanks for the patience!
